stratum/lib/p4runtime/sdn_controller_manager.cc

```diff
@@ -0,0 +1,445 @@
+    if (connection->GetRoleName() == role_name &&
+        connection->GetElectionId() == election_id) {
+      return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
+                          "Election ID is already used by another connection "
+                          "with the same role.");
+    }
+  }
+  return grpc::Status::OK;
+}
+
+grpc::Status VerifyElectionIdIsActive(
+    const absl::optional<std::string>& role_name,
+    const absl::optional<absl::uint128>& election_id,
+    absl::Span<const SdnConnection* const> active_connections) {
+  for (const auto& connection : active_connections) {
+    if (connection->GetRoleName() == role_name &&
+        connection->GetElectionId() == election_id) {
+      return grpc::Status::OK;
+    }
+  }
+  return grpc::Status(grpc::StatusCode::PERMISSION_DENIED,
+                      "Election ID is not active for the role.");
+}
+
+}  // namespace
+
+void SdnConnection::SetElectionId(const absl::optional<absl::uint128>& id) {
+  election_id_ = id;
+}
+
+absl::optional<absl::uint128> SdnConnection::GetElectionId() const {
+  return election_id_;
+}
+
+void SdnConnection::SetRoleName(const absl::optional<std::string>& name) {
+  role_name_ = name;
+}
+
+absl::optional<std::string> SdnConnection::GetRoleName() const {
+  return role_name_;
+}
+
+std::string SdnConnection::GetName() const {
+  return absl::StrCat("(role_name: ", PrettyPrintRoleName(role_name_),
+                      ", election_id: ", PrettyPrintElectionId(election_id_),
+                      ", uri: ", grpc_context_->peer(), ")");
+}
+
+void SdnConnection::SendStreamMessageResponse(
+    const p4::v1::StreamMessageResponse& response) {
+  VLOG(2) << "Sending response: " << response.ShortDebugString();
+  if (!grpc_stream_->Write(response)) {
+    LOG(ERROR) << "Could not send arbitration update response to gRPC context '"
+               << grpc_context_ << "': " << response.ShortDebugString();
+  }
+}
+
+grpc::Status SdnControllerManager::HandleArbitrationUpdate(
+    const p4::v1::MasterArbitrationUpdate& update, SdnConnection* controller) {
+  absl::MutexLock l(&lock_);
+
+  // If the role name is not set then we assume the connection is a 'root'
+  // connection.
+  absl::optional<std::string> role_name;
+#if P4RT_HAS_ROLES
+  if (update.has_role() && !update.role().name().empty()) {
+    role_name = update.role().name();
+  }
+#endif
+
+  const auto old_election_id_for_connection = controller->GetElectionId();
+  absl::optional<absl::uint128> new_election_id_for_connection;
+  if (update.has_election_id()) {
+    new_election_id_for_connection = absl::MakeUint128(
+        update.election_id().high(), update.election_id().low());
+  }
+
+  const bool new_connection = !controller->IsInitialized();
+
+  if (new_connection) {
+    // First arbitration message sent by this controller.
+
+    // Verify the request's device ID is being sent to the correct device.
+    if (update.device_id() != device_id_) {
```

Comment:
Please correct me if I am wrong
After the client creates a new gRPC stream, it will send a new arbitration update message to the server, which includes a `device_id` in the message and we store this data in `node_id` in `p4_service`([line 573](https://github.com/stratum/stratum/blob/82b2ff66659fc2324f213e3b3b10da081b4a704b/stratum/hal/lib/common/p4_service.cc#L573)).
Everytime when the client sends a `StreamMessageRequest` we will check if the `device_id` changed(`p4_service` [line 574](https://github.com/stratum/stratum/blob/82b2ff66659fc2324f213e3b3b10da081b4a704b/stratum/hal/lib/common/p4_service.cc#L574))
Does this mean we don't need to check the id again in this function? Or we can simplify the check in the `p4_service`?

Reply:
Also, I was thinking if this condition will always be false since we always get the `sdn_controller_manager` based on the `device_id`/`node_id` ([line 718](https://github.com/stratum/stratum/blob/82b2ff66659fc2324f213e3b3b10da081b4a704b/stratum/hal/lib/common/p4_service.cc#L684))

Reply:
comments also apply to line 161

Reply:
There is a duplicate check happening, true. I think the sdn_controller should be as simple and safe to use as possible, hence should be in charge of doing the check. Then we can probably remove it from p4 service.

Reply:
done